File: app/analytics/metrics.py
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from .models import EffortMetrics, ForceSample, KG_TO_N


@dataclass(slots=True)
class AnalysisConfig:
    tut_threshold_kg: float = 0.5
    hold_peak_fraction: float = 0.9
    stabilization_shift_threshold: float = 0.8  # sum(|dp_i/dt|), where p_i in [0,1]
    stabilization_hold_ms: int = 250


def _interp_at_ms(times_ms: np.ndarray, values: np.ndarray, target_ms: float) -> float:
    if len(times_ms) == 0:
        return 0.0
    if target_ms <= times_ms[0]:
        return float(values[0])
    if target_ms >= times_ms[-1]:
        return float(values[-1])

    idx = int(np.searchsorted(times_ms, target_ms, side="left"))
    if idx <= 0:
        return float(values[0])
    t0, t1 = float(times_ms[idx - 1]), float(times_ms[idx])
    v0, v1 = float(values[idx - 1]), float(values[idx])
    if t1 <= t0:
        return v1
    frac = (target_ms - t0) / (t1 - t0)
    return (v0 * (1.0 - frac)) + (v1 * frac)


def _duration_above_threshold_s(times_ms: np.ndarray, values: np.ndarray, threshold: float) -> float:
    if len(times_ms) < 2:
        return 0.0
    acc_ms = 0.0
    for i in range(len(values) - 1):
        if values[i] >= threshold:
            acc_ms += max(0.0, times_ms[i + 1] - times_ms[i])
    return acc_ms / 1000.0


def _percentages(fingers_kg: np.ndarray, total_kg: np.ndarray) -> np.ndarray:
    out = np.zeros_like(fingers_kg)
    valid = total_kg > 1e-9
    if np.any(valid):
        out[valid] = fingers_kg[valid] / total_kg[valid, None]
    return out


def _load_shift_rate_series(times_ms: np.ndarray, pct: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    if len(times_ms) < 2:
        return np.array([], dtype=float), np.array([], dtype=float)

    dt_s = np.diff(times_ms) / 1000.0
    dt_s[dt_s <= 0.0] = 1e-6
    dp = np.abs(np.diff(pct, axis=0))
    shift = np.sum(dp, axis=1) / dt_s
    t_ref = times_ms[1:]
    return t_ref, shift


def _stabilization_time_s(
    times_ms: np.ndarray,
    pct: np.ndarray,
    start_t_ms: int,
    threshold: float,
    hold_ms: int,
) -> float | None:
    t_ref, shift = _load_shift_rate_series(times_ms, pct)
    if len(shift) == 0:
        return None

    candidate_start: float | None = None
    accum_ms = 0.0

    for i, rate in enumerate(shift):
        if rate < threshold:
            if candidate_start is None:
                candidate_start = float(t_ref[i])
                accum_ms = 0.0
            if i > 0:
                accum_ms += max(0.0, float(t_ref[i] - t_ref[i - 1]))
            if accum_ms >= hold_ms and candidate_start is not None:
                return max(0.0, (candidate_start - start_t_ms) / 1000.0)
        else:
            candidate_start = None
            accum_ms = 0.0
    return None


def _rfd_kg_s(times_ms: np.ndarray, total_kg: np.ndarray, t0_ms: int, window_ms: int) -> float:
    if len(times_ms) == 0:
        return 0.0
    f0 = _interp_at_ms(times_ms, total_kg, float(t0_ms))
    f1 = _interp_at_ms(times_ms, total_kg, float(t0_ms + window_ms))
    return (f1 - f0) / (window_ms / 1000.0)


def analyze_effort_samples(
    effort_samples: list[ForceSample],
    effort_id: int,
    config: AnalysisConfig,
) -> EffortMetrics:
    if len(effort_samples) < 2:
        s = effort_samples[0]
        return EffortMetrics(
            effort_id=effort_id,
            start_t_ms=s.t_ms,
            end_t_ms=s.t_ms,
            duration_s=0.0,
            peak_total_kg=s.total_kg,
            peak_per_finger_kg=s.kg,
            time_to_peak_s=0.0,
            rfd_100_kg_s=0.0,
            rfd_200_kg_s=0.0,
            rfd_100_n_s=0.0,
            rfd_200_n_s=0.0,
            avg_total_kg=s.total_kg,
            tut_s=0.0,
            distribution_drift_per_s=0.0,
            steadiness_total_kg=0.0,
            steadiness_per_finger_kg=(0.0, 0.0, 0.0, 0.0),
            finger_imbalance_index=0.0,
            load_variation_cv=0.0,
            dominant_switch_count=0,
            load_shift_rate=0.0,
            stabilization_time_s=None,
            ring_pinky_share=0.0,
            hold_start_t_ms=s.t_ms,
            hold_end_t_ms=s.t_ms,
            detail_t_ms=[0],
            detail_total_kg=[s.total_kg],
            detail_finger_kg=[s.kg],
            detail_finger_pct=[(0.0, 0.0, 0.0, 0.0)],
        )

    times_ms = np.array([s.t_ms for s in effort_samples], dtype=float)
    fingers = np.array([s.kg for s in effort_samples], dtype=float)
    total = np.sum(fingers, axis=1)

    start_t_ms = int(times_ms[0])
    end_t_ms = int(times_ms[-1])
    duration_s = max(0.0, (end_t_ms - start_t_ms) / 1000.0)

    peak_idx = int(np.argmax(total))
    peak_total = float(total[peak_idx])
    peak_fingers = tuple(float(v) for v in fingers[peak_idx])
    time_to_peak_s = max(0.0, (float(times_ms[peak_idx]) - start_t_ms) / 1000.0)

    rfd_100 = _rfd_kg_s(times_ms, total, start_t_ms, 100)
    rfd_200 = _rfd_kg_s(times_ms, total, start_t_ms, 200)

    hold_threshold = config.hold_peak_fraction * peak_total
    hold_idx = 0
    for i in range(len(total)):
        if total[i] >= hold_threshold:
            hold_idx = i
            break

    hold_times = times_ms[hold_idx:]
    hold_fingers = fingers[hold_idx:]
    hold_total = total[hold_idx:]

    pct_all = _percentages(fingers, total)
    pct_hold = _percentages(hold_fingers, hold_total)

    hold_start_t_ms = int(times_ms[hold_idx])
    hold_end_t_ms = int(times_ms[-1])

    avg_total_kg = float(np.mean(hold_total)) if len(hold_total) else 0.0
    tut_s = _duration_above_threshold_s(times_ms, total, config.tut_threshold_kg)

    # Distribution drift per second during hold
    if len(hold_times) >= 2:
        _, hold_shift_series = _load_shift_rate_series(hold_times, pct_hold)
        distribution_drift = float(np.mean(hold_shift_series)) if len(hold_shift_series) else 0.0
    else:
        distribution_drift = 0.0

    steadiness_total = float(np.std(hold_total)) if len(hold_total) else 0.0
    steadiness_finger = (
        float(np.std(hold_fingers[:, 0])) if len(hold_fingers) else 0.0,
        float(np.std(hold_fingers[:, 1])) if len(hold_fingers) else 0.0,
        float(np.std(hold_fingers[:, 2])) if len(hold_fingers) else 0.0,
        float(np.std(hold_fingers[:, 3])) if len(hold_fingers) else 0.0,
    )

    # Imbalance index = mean stddev over finger percentages (percentage points)
    imbalance = float(np.mean(np.std(pct_hold * 100.0, axis=1))) if len(pct_hold) else 0.0

    # Load variation CV = avg CV over fingers during hold
    cvs = []
    for i in range(4):
        ch = hold_fingers[:, i] if len(hold_fingers) else np.array([], dtype=float)
        if len(ch) == 0:
            cvs.append(0.0)
            continue
        mean_abs = float(np.mean(np.abs(ch)))
        if mean_abs < 1e-9:
            cvs.append(0.0)
        else:
            cvs.append(float(np.std(ch) / mean_abs))
    load_variation_cv = float(np.mean(cvs)) if cvs else 0.0

    dom_idx = np.argmax(pct_all, axis=1) if len(pct_all) else np.array([], dtype=int)
    switch_count = int(np.sum(dom_idx[1:] != dom_idx[:-1])) if len(dom_idx) > 1 else 0

    _, shift_series = _load_shift_rate_series(times_ms, pct_all)
    load_shift_rate = float(np.mean(shift_series)) if len(shift_series) else 0.0

    stabilization_time = _stabilization_time_s(
        times_ms,
        pct_all,
        start_t_ms,
        threshold=config.stabilization_shift_threshold,
        hold_ms=config.stabilization_hold_ms,
    )

    valid_hold = hold_total > 1e-9
    if np.any(valid_hold):
        ring_pinky_share = float(np.mean((hold_fingers[valid_hold, 2] + hold_fingers[valid_hold, 3]) / hold_total[valid_hold]))
    else:
        ring_pinky_share = 0.0

    rel_t = [int(t - start_t_ms) for t in times_ms]
    detail_total = [float(v) for v in total]
    detail_fingers = [tuple(float(v) for v in row) for row in fingers]
    detail_pct = [tuple(float(v * 100.0) for v in row) for row in pct_all]

    return EffortMetrics(
        effort_id=effort_id,
        start_t_ms=start_t_ms,
        end_t_ms=end_t_ms,
        duration_s=duration_s,
        peak_total_kg=peak_total,
        peak_per_finger_kg=peak_fingers,
        time_to_peak_s=time_to_peak_s,
        rfd_100_kg_s=rfd_100,
        rfd_200_kg_s=rfd_200,
        rfd_100_n_s=rfd_100 * KG_TO_N,
        rfd_200_n_s=rfd_200 * KG_TO_N,
        avg_total_kg=avg_total_kg,
        tut_s=tut_s,
        distribution_drift_per_s=distribution_drift,
        steadiness_total_kg=steadiness_total,
        steadiness_per_finger_kg=steadiness_finger,
        finger_imbalance_index=imbalance,
        load_variation_cv=load_variation_cv,
        dominant_switch_count=switch_count,
        load_shift_rate=load_shift_rate,
        stabilization_time_s=stabilization_time,
        ring_pinky_share=ring_pinky_share,
        hold_start_t_ms=hold_start_t_ms,
        hold_end_t_ms=hold_end_t_ms,
        detail_t_ms=rel_t,
        detail_total_kg=detail_total,
        detail_finger_kg=detail_fingers,
        detail_finger_pct=detail_pct,
    )

File: packages/core/src/metrics.ts
import type { EffortMetrics, Finger4, ForceSample } from './types.ts';
import { KG_TO_N, totalKg } from './types.ts';

export interface AnalysisConfig {
  tutThresholdKg: number;
  holdPeakFraction: number;
  stabilizationShiftThreshold: number;
  stabilizationHoldMs: number;
}

export const DEFAULT_ANALYSIS_CONFIG: AnalysisConfig = {
  tutThresholdKg: 0.5,
  holdPeakFraction: 0.9,
  stabilizationShiftThreshold: 0.8,
  stabilizationHoldMs: 250,
};

function interpAtMs(timesMs: number[], values: number[], targetMs: number): number {
  if (timesMs.length === 0) return 0;
  if (targetMs <= timesMs[0]) return values[0];
  if (targetMs >= timesMs[timesMs.length - 1]) return values[values.length - 1];

  let idx = 0;
  for (let i = 0; i < timesMs.length; i += 1) {
    if (timesMs[i] >= targetMs) {
      idx = i;
      break;
    }
  }

  if (idx <= 0) return values[0];

  const t0 = timesMs[idx - 1];
  const t1 = timesMs[idx];
  const v0 = values[idx - 1];
  const v1 = values[idx];
  if (t1 <= t0) return v1;
  const frac = (targetMs - t0) / (t1 - t0);
  return v0 * (1 - frac) + v1 * frac;
}

function rfdKgS(timesMs: number[], totalValues: number[], t0Ms: number, windowMs: number): number {
  if (timesMs.length === 0) return 0;
  const f0 = interpAtMs(timesMs, totalValues, t0Ms);
  const f1 = interpAtMs(timesMs, totalValues, t0Ms + windowMs);
  return (f1 - f0) / (windowMs / 1000);
}

function durationAboveThreshold(timesMs: number[], values: number[], threshold: number): number {
  if (timesMs.length < 2) return 0;
  let accMs = 0;
  for (let i = 0; i < values.length - 1; i += 1) {
    if (values[i] >= threshold) {
      accMs += Math.max(0, timesMs[i + 1] - timesMs[i]);
    }
  }
  return accMs / 1000;
}

function mean(values: number[]): number {
  if (values.length === 0) return 0;
  let total = 0;
  for (const value of values) total += value;
  return total / values.length;
}

function stddev(values: number[]): number {
  if (values.length < 2) return 0;
  const m = mean(values);
  let sumSq = 0;
  for (const value of values) {
    const delta = value - m;
    sumSq += delta * delta;
  }
  return Math.sqrt(sumSq / values.length);
}

function percentages(fingersKg: Finger4[], totalsKg: number[]): Finger4[] {
  return fingersKg.map((fingerRow, index) => {
    const total = totalsKg[index];
    if (total <= 1e-9) return [0, 0, 0, 0];
    return [
      fingerRow[0] / total,
      fingerRow[1] / total,
      fingerRow[2] / total,
      fingerRow[3] / total,
    ];
  });
}

function loadShiftRateSeries(timesMs: number[], pctSeries: Finger4[]): { times: number[]; shift: number[] } {
  if (timesMs.length < 2) return { times: [], shift: [] };

  const shift: number[] = [];
  const times: number[] = [];

  for (let i = 1; i < timesMs.length; i += 1) {
    const dtSeconds = Math.max(1e-6, (timesMs[i] - timesMs[i - 1]) / 1000);
    const dp =
      Math.abs(pctSeries[i][0] - pctSeries[i - 1][0]) +
      Math.abs(pctSeries[i][1] - pctSeries[i - 1][1]) +
      Math.abs(pctSeries[i][2] - pctSeries[i - 1][2]) +
      Math.abs(pctSeries[i][3] - pctSeries[i - 1][3]);
    shift.push(dp / dtSeconds);
    times.push(timesMs[i]);
  }

  return { times, shift };
}

function stabilizationTimeS(
  timesMs: number[],
  pctSeries: Finger4[],
  startTMs: number,
  threshold: number,
  holdMs: number,
): number | null {
  const { times, shift } = loadShiftRateSeries(timesMs, pctSeries);
  if (shift.length === 0) return null;

  let candidateStart: number | null = null;
  let accumMs = 0;

  for (let i = 0; i < shift.length; i += 1) {
    if (shift[i] < threshold) {
      if (candidateStart === null) {
        candidateStart = times[i];
        accumMs = 0;
      }
      if (i > 0) {
        accumMs += Math.max(0, times[i] - times[i - 1]);
      }
      if (accumMs >= holdMs && candidateStart !== null) {
        return Math.max(0, (candidateStart - startTMs) / 1000);
      }
    } else {
      candidateStart = null;
      accumMs = 0;
    }
  }

  return null;
}

function zeroFinger4(): Finger4 {
  return [0, 0, 0, 0];
}

function hasPerFingerData(samples: ForceSample[]): samples is Array<ForceSample & { kg: Finger4 }> {
  return samples.every(sample => sample.kg !== null);
}

function analyzeTotalOnlyEffort(
  effortSamples: ForceSample[],
  effortId: number,
  config: AnalysisConfig,
): EffortMetrics {
  const timesMs = effortSamples.map(sample => sample.tMs);
  const totals = effortSamples.map(sample => totalKg(sample));
  const startTMs = timesMs[0];
  const endTMs = timesMs[timesMs.length - 1];
  const durationS = Math.max(0, (endTMs - startTMs) / 1000);

  let peakIdx = 0;
  let peakTotal = totals[0] ?? 0;
  for (let i = 1; i < totals.length; i += 1) {
    if (totals[i] > peakTotal) {
      peakTotal = totals[i];
      peakIdx = i;
    }
  }

  const holdThreshold = config.holdPeakFraction * peakTotal;
  let holdIdx = 0;
  for (let i = 0; i < totals.length; i += 1) {
    if (totals[i] >= holdThreshold) {
      holdIdx = i;
      break;
    }
  }

  const holdTotalsKg = totals.slice(holdIdx);

  return {
    effortId,
    startTMs,
    endTMs,
    durationS,
    peakTotalKg: peakTotal,
    peakPerFingerKg: null,
    timeToPeakS: Math.max(0, (timesMs[peakIdx] - startTMs) / 1000),
    rfd100KgS: rfdKgS(timesMs, totals, startTMs, 100),
    rfd200KgS: rfdKgS(timesMs, totals, startTMs, 200),
    rfd100NS: rfdKgS(timesMs, totals, startTMs, 100) * KG_TO_N,
    rfd200NS: rfdKgS(timesMs, totals, startTMs, 200) * KG_TO_N,
    avgTotalKg: holdTotalsKg.length > 0 ? mean(holdTotalsKg) : 0,
    tutS: durationAboveThreshold(timesMs, totals, config.tutThresholdKg),
    distributionDriftPerS: null,
    steadinessTotalKg: holdTotalsKg.length > 0 ? stddev(holdTotalsKg) : 0,
    steadinessPerFingerKg: null,
    fingerImbalanceIndex: null,
    loadVariationCv: null,
    dominantSwitchCount: null,
    loadShiftRate: null,
    stabilizationTimeS: null,
    ringPinkyShare: null,
    holdStartTMs: timesMs[holdIdx] ?? startTMs,
    holdEndTMs: endTMs,
    detailTMs: timesMs.map(time => time - startTMs),
    detailTotalKg: [...totals],
    detailFingerKg: null,
    detailFingerPct: null,
  };
}

export function analyzeEffortSamples(
  effortSamples: ForceSample[],
  effortId: number,
  config: AnalysisConfig = DEFAULT_ANALYSIS_CONFIG,
): EffortMetrics {
  if (effortSamples.length === 0) {
    return {
      effortId,
      startTMs: 0,
      endTMs: 0,
      durationS: 0,
      peakTotalKg: 0,
      peakPerFingerKg: null,
      timeToPeakS: 0,
      rfd100KgS: 0,
      rfd200KgS: 0,
      rfd100NS: 0,
      rfd200NS: 0,
      avgTotalKg: 0,
      tutS: 0,
      distributionDriftPerS: null,
      steadinessTotalKg: 0,
      steadinessPerFingerKg: null,
      fingerImbalanceIndex: null,
      loadVariationCv: null,
      dominantSwitchCount: null,
      loadShiftRate: null,
      stabilizationTimeS: null,
      ringPinkyShare: null,
      holdStartTMs: 0,
      holdEndTMs: 0,
      detailTMs: [],
      detailTotalKg: [],
      detailFingerKg: null,
      detailFingerPct: null,
    };
  }

  if (effortSamples.length < 2) {
    const sample = effortSamples[0];
    return {
      effortId,
      startTMs: sample.tMs,
      endTMs: sample.tMs,
      durationS: 0,
      peakTotalKg: totalKg(sample),
      peakPerFingerKg: sample.kg,
      timeToPeakS: 0,
      rfd100KgS: 0,
      rfd200KgS: 0,
      rfd100NS: 0,
      rfd200NS: 0,
      avgTotalKg: totalKg(sample),
      tutS: 0,
      distributionDriftPerS: sample.kg ? 0 : null,
      steadinessTotalKg: 0,
      steadinessPerFingerKg: sample.kg ? zeroFinger4() : null,
      fingerImbalanceIndex: sample.kg ? 0 : null,
      loadVariationCv: sample.kg ? 0 : null,
      dominantSwitchCount: sample.kg ? 0 : null,
      loadShiftRate: sample.kg ? 0 : null,
      stabilizationTimeS: null,
      ringPinkyShare: sample.kg ? 0 : null,
      holdStartTMs: sample.tMs,
      holdEndTMs: sample.tMs,
      detailTMs: [0],
      detailTotalKg: [totalKg(sample)],
      detailFingerKg: sample.kg ? [sample.kg] : null,
      detailFingerPct: sample.kg ? [zeroFinger4()] : null,
    };
  }

  if (!hasPerFingerData(effortSamples)) {
    return analyzeTotalOnlyEffort(effortSamples, effortId, config);
  }

  const timesMs = effortSamples.map(sample => sample.tMs);
  const fingers = effortSamples.map(sample => sample.kg);
  const totals = fingers.map(finger => finger[0] + finger[1] + finger[2] + finger[3]);

  const startTMs = timesMs[0];
  const endTMs = timesMs[timesMs.length - 1];
  const durationS = Math.max(0, (endTMs - startTMs) / 1000);

  let peakIdx = 0;
  let peakTotal = totals[0];
  for (let i = 1; i < totals.length; i += 1) {
    if (totals[i] > peakTotal) {
      peakTotal = totals[i];
      peakIdx = i;
    }
  }

  const peakPerFingerKg = fingers[peakIdx];
  const timeToPeakS = Math.max(0, (timesMs[peakIdx] - startTMs) / 1000);

  const rfd100KgS = rfdKgS(timesMs, totals, startTMs, 100);
  const rfd200KgS = rfdKgS(timesMs, totals, startTMs, 200);

  const holdThreshold = config.holdPeakFraction * peakTotal;
  let holdIdx = 0;
  for (let i = 0; i < totals.length; i += 1) {
    if (totals[i] >= holdThreshold) {
      holdIdx = i;
      break;
    }
  }

  const holdTimesMs = timesMs.slice(holdIdx);
  const holdFingersKg = fingers.slice(holdIdx);
  const holdTotalsKg = totals.slice(holdIdx);

  const allPct = percentages(fingers, totals);
  const holdPct = percentages(holdFingersKg, holdTotalsKg);

  const holdStartTMs = timesMs[holdIdx];
  const holdEndTMs = timesMs[timesMs.length - 1];
  const avgTotalKg = holdTotalsKg.length > 0 ? mean(holdTotalsKg) : 0;
  const tutS = durationAboveThreshold(timesMs, totals, config.tutThresholdKg);

  let distributionDriftPerS = 0;
  if (holdTimesMs.length >= 2) {
    const holdShift = loadShiftRateSeries(holdTimesMs, holdPct).shift;
    distributionDriftPerS = holdShift.length > 0 ? mean(holdShift) : 0;
  }

  const steadinessTotalKg = holdTotalsKg.length > 0 ? stddev(holdTotalsKg) : 0;
  const steadinessPerFingerKg: Finger4 = holdFingersKg.length > 0
    ? [
        stddev(holdFingersKg.map(finger => finger[0])),
        stddev(holdFingersKg.map(finger => finger[1])),
        stddev(holdFingersKg.map(finger => finger[2])),
        stddev(holdFingersKg.map(finger => finger[3])),
      ]
    : zeroFinger4();

  const fingerImbalanceIndex = holdPct.length > 0
    ? mean(holdPct.map(row => stddev(row.map(value => value * 100))))
    : 0;

  const loadVariationValues: number[] = [];
  for (let fingerIdx = 0; fingerIdx < 4; fingerIdx += 1) {
    const series = holdFingersKg.map(finger => finger[fingerIdx]);
    if (series.length === 0) {
      loadVariationValues.push(0);
      continue;
    }
    const meanAbs = mean(series.map(value => Math.abs(value)));
    if (meanAbs < 1e-9) {
      loadVariationValues.push(0);
      continue;
    }
    loadVariationValues.push(stddev(series) / meanAbs);
  }
  const loadVariationCv = loadVariationValues.length > 0 ? mean(loadVariationValues) : 0;

  let dominantSwitchCount = 0;
  let previousDominant = -1;
  for (const row of allPct) {
    let dominant = 0;
    for (let i = 1; i < row.length; i += 1) {
      if (row[i] > row[dominant]) dominant = i;
    }
    if (previousDominant >= 0 && dominant !== previousDominant) {
      dominantSwitchCount += 1;
    }
    previousDominant = dominant;
  }

  const loadShiftRateSeriesAll = loadShiftRateSeries(timesMs, allPct);
  const loadShiftRate = loadShiftRateSeriesAll.shift.length > 0 ? mean(loadShiftRateSeriesAll.shift) : 0;

  const stabilizationTimeSValue = stabilizationTimeS(
    timesMs,
    allPct,
    startTMs,
    config.stabilizationShiftThreshold,
    config.stabilizationHoldMs,
  );

  const validHoldShares = holdTotalsKg
    .map((total, index) => ({ total, fingers: holdFingersKg[index] }))
    .filter(entry => entry.total > 1e-9)
    .map(entry => (entry.fingers[2] + entry.fingers[3]) / entry.total);
  const ringPinkyShare = validHoldShares.length > 0 ? mean(validHoldShares) : 0;

  const detailTMs = timesMs.map(time => time - startTMs);
  const detailFingerPct = allPct.map(row => [
    row[0] * 100,
    row[1] * 100,
    row[2] * 100,
    row[3] * 100,
  ] as Finger4);

  return {
    effortId,
    startTMs,
    endTMs,
    durationS,
    peakTotalKg: peakTotal,
    peakPerFingerKg,
    timeToPeakS,
    rfd100KgS,
    rfd200KgS,
    rfd100NS: rfd100KgS * KG_TO_N,
    rfd200NS: rfd200KgS * KG_TO_N,
    avgTotalKg,
    tutS,
    distributionDriftPerS,
    steadinessTotalKg,
    steadinessPerFingerKg,
    fingerImbalanceIndex,
    loadVariationCv,
    dominantSwitchCount,
    loadShiftRate,
    stabilizationTimeS: stabilizationTimeSValue,
    ringPinkyShare,
    holdStartTMs,
    holdEndTMs,
    detailTMs,
    detailTotalKg: [...totals],
    detailFingerKg: fingers.map(finger => [...finger] as Finger4),
    detailFingerPct,
  };
}
